### PR TITLE
[feat] proxmox_snap: snapshot containers with configured mountpoints

### PR DESCRIPTION
**Pull request by nxet**
##### SUMMARY
This PR adds a new parameter `unbind` to the `proxmox_snap` module, which allows to take snapshots of containers with configured mount points, bypassing the Proxmox limit by *temporarily* disabling the configuration before the snapshot.  

There is also a small improvement in how the API tasks are checked.  
A new `api_task_ok` helper is provided in the `module_utils.proxmox.ProxmoxAnsible` class, and used across the proxmox modules to standardize how pending tasks are checked.  

##### ISSUE TYPE
- Feature Pull Request

##### COMPONENT NAME
community.general.proxmox_snap

##### ADDITIONAL INFORMATION
`proxmox_snap`  

The mountpoint-specific config API endpoint `mp[n]` is unfortunately (from an implementation pov, of course it's only logical) restricted to `root@pam` authenticating with password. Tokens will not work, at least I couldn't get them to work. More info [here, PUT tab](https://pve.proxmox.com/pve-docs/api-viewer/#/nodes/{node}/lxc/{vmid}/config).  
This limit is not enforced when deleting the mounts config though, which means bad credentials would allow to remove the mount points **BUT fail at reconfig**, leaving the container in a misconfigured state. This behavior is properly commented in-code, documented, and the implementation checks if the user provided the correct authentication parameters, eventually failing before running any task.

Mainly to improve readability, two new helpers `start_instance` and `shutdown_instance` have been introduced, as well as internal helpers to get and manipulate the mountpoints configuration (`_container_mp_{get,disable,restore}`).  

Changes to the container's mountpoints config require shutting down the machine. The state of the container (`running` or not) is stored before applying any changes, and restored after the snapshot if necessary. This behavior is also properly documented.

Finally the new logic tries to gracefully handle timeouts during the snapshot, by restoring the original config and state regardless of the successful execution of the API task.

`api_task_ok`
There could be some room for improvement, providing a `wait_api_task` helper to also abstract the `while timeout:` blocks across the modules, but it would have been out of scope for this PR. 

##### Test Playbook
```yml
---

#
# requirements:
#   - create `/tmp/foo` and `/tmp/bar` on the target proxmox node
#

- hosts: localhost

  vars_prompt:
    - name: proxmox_root_pass
      prompt: "Enter Proxmox password"
      private: true

  tasks:

    - name: Create new container with minimal options
      community.general.proxmox:
        vmid: 100
        node: px10
        api_user: root@pam
        api_password: "{{ proxmox_root_pass }}"
        api_host: px10.example.com
        password: 123456
        hostname: test
        ostemplate: 'local:vztmpl/alpine-3.16-default_20220622_amd64.tar.xz'
        mounts:
          mp0: /tmp/foo,mp=/tmp/foo
          mp1: /tmp/bar,mp=/tmp/bar

    - name:
      community.general.proxmox_snap:
        vmid: 100
        api_user: root@pam
        api_password: "{{ proxmox_root_pass }}"
        api_host: px10.example.com
        state: present
        snapname: "test_{{ '%Y%m%d%H%M' | strftime }}"
        unbind: true

...
```

### CODE REVIEW

File: plugins/module_utils/proxmox.py
Comment (felixfontein) [inline plugins/module_utils/proxmox.py:145]: Why not simply:
```suggestion
        return status['status'] == 'stopped' and status['exitstatus'] == 'OK'
```
?

File: plugins/modules/cloud/misc/proxmox_snap.py
Comment (felixfontein) [inline plugins/modules/cloud/misc/proxmox_snap.py:50]:
```suggestion
    type: bool
    version_added: 5.6.0
```

File: plugins/modules/cloud/misc/proxmox_snap.py
Comment (felixfontein) [inline plugins/modules/cloud/misc/proxmox_snap.py]:
```suggestion
      - Due to restrictions in the Proxmox API this option can only be used authenticating as C(root@pam) with I(api_password) set, API tokens do not work either. See U(https://pve.proxmox.com/pve-docs/api-viewer/#/nodes/{node}/lxc/{vmid}/config) (PUT tab) for more details.
```

**Review by felixfontein — COMMENTED** [review]
Thanks for your contribution! I've added some first comments. Can you please add a [changelog fragment](https://docs.ansible.com/ansible/devel/community/collection_development_process.html#creating-a-changelog-fragment)? Thanks.

File: plugins/module_utils/proxmox.py
Comment (nxet) [reply plugins/module_utils/proxmox.py:145]: very very nice :+1:

**Comment by nxet** [comment]
The last two commits address the remarks above, specifically:
- add changelog fragment, for both the `unbind` param and the `api_task_ok` helper
- improve the `api_task_ok` logic
- fix tag error in documentation
- add `version_added` tag to the `unbind` param docs
- fix formatting errors

File: changelogs/fragments/5274-proxmox-snap-container-with-mountpoints.yml
Comment (felixfontein) [inline changelogs/fragments/5274-proxmox-snap-container-with-mountpoints.yml:2]:
```suggestion
  - proxmox_snap - add ``unbind`` param to support snapshotting containers with configured mountpoints (https://github.com/ansible-collections/community.general/pull/5274).
```

File: changelogs/fragments/5274-proxmox-snap-container-with-mountpoints.yml
Comment (felixfontein) [inline changelogs/fragments/5274-proxmox-snap-container-with-mountpoints.yml:3]:
```suggestion
  - proxmox module utils, the proxmox* modules - add ``api_task_ok`` helper to standardize API task status checks across all proxmox modules (https://github.com/ansible-collections/community.general/pull/5274).
```

**Comment by nxet** [comment]
Great to see all tests are passing, looking forward to the release!

The commit history obviously requires a rebase, will you take care of it when merging?
Should I do it myself and push the changes?
As you can see I'm not very active on Github, and I'm not sure about the required next steps. 
Thanks for your understanding, and for all the suggestions during code review.

Just wanted to add that I'm *super happy* to be contributing to Ansible!
I'm beyond honored to have my small PR merged to such a massive software suite.

File: plugins/modules/cloud/misc/proxmox_snap.py
Comment (felixfontein) [inline plugins/modules/cloud/misc/proxmox_snap.py:51]: Unfortunately 5.6.0 is already out, so this needs to be updated to 5.7.0:
```suggestion
    version_added: 5.7.0
```

**Review by felixfontein — COMMENTED** [review]
Besides the following point, this looks good to me. I cannot really test this, but the code changes look good to me (i.e. they don't seem to break something that already worked) and since none of the maintainers reacted so far, I guess nobody objects. I'll merge this in a couple of days (assuming the below modification has been done).

**Comment by felixfontein** [comment]
> The commit history obviously requires a rebase, will you take care of it when merging? Should I do it myself and push the changes?

We're always squashing on merge, so a single commit will be added to the `main` branch for every PR.

> As you can see I'm not very active on Github, and I'm not sure about the required next steps. Thanks for your understanding, and for all the suggestions during code review.

I want to give the maintainers of this module some time to also take a look, but if they don't (or don't feel to write a comment) I'll just merge it (at least in cases like this where I'm pretty sure it won't break anything).

> Just wanted to add that I'm _super happy_ to be contributing to Ansible! I'm beyond honored to have my small PR merged to such a massive software suite.

And we're very glad that you decided to contribute to it :) This part of Ansible is purely community driven, so without folks who spend time contributing (either by creating issues and PRs, or by reviewing / triaging / ... them) like you nothing would happen!

**Comment by nxet** [comment]
Thanks again for your kindness and patience, both with the code review and helping me better understand.

I agree, and it would definitely be reassuring to also get the green light from the maintainers. There is absolutely no rush to merge, so maybe we can let this sit for a couple extra days?
Most of all it would be really nice to have somebody else run the test playbook, just to make sure this is not a "...but it works on my machine" kind of situation.

**Review by felixfontein — APPROVED** [review]

**Comment by felixfontein** [comment]
I think we've waited long enough...

**Comment by felixfontein** [comment]
@nxet thanks for your contribution!

**Pull request merged**
